src/Corax/IndexWriter.cs

```diff
@@ -1829,177 +1838,426 @@ private void InsertSpatialField(Tree entriesToSpatialTree, IndexedField indexedF
             }
         }
 
-        private void InsertTextualField(Tree entriesToTermsTree, IndexedField indexedField, Span<byte> tmpBuf, ref Slice[] sortedTermsBuffer)
+        
+        private class TextualFieldBuffers : IDisposable
         {
-            var fieldTree = _fieldsTree.CompactTreeFor(indexedField.Name);
+            private readonly IndexWriter _parent;
+            public const int BatchSize = 1024;
+
+            private Slice[] _sortedTerms;
+            private int[] _termIndexes;
             
-            var currentFieldTerms = indexedField.Textual;
-            int termsCount = currentFieldTerms.Count;
+            public CompactTree.CompactKeyLookup[] Keys;
+            public int[] PageOffsets;
+            public long[] PostListIds;
+            private int[] _entriesOffsets;
 
-            ClearEntriesForTerm();
+            public void PrepareTerms(IndexedField field, out Span<Slice> terms, out Span<int> indexes)
+            {
+                int termsCount = field.Textual.Count;
+                if (_sortedTerms == null || _sortedTerms.Length < termsCount)
+                {
+                    if (_sortedTerms != null)
+                    {
+                        ArrayPool<Slice>.Shared.Return(_sortedTerms);
+                        ArrayPool<int>.Shared.Return(_termIndexes);
+                    }
+                    _sortedTerms = ArrayPool<Slice>.Shared.Rent(termsCount);
+                    _termIndexes = ArrayPool<int>.Shared.Rent(termsCount);
+                }
 
-            if (sortedTermsBuffer.Length < termsCount)
+                int idx = 0;
+                foreach (var (k,v) in field.Textual)
+                {
+                    _sortedTerms[idx] = k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+
+                terms = new Span<Slice>(_sortedTerms, 0, termsCount);
+                indexes = new Span<int>(_termIndexes, 0, termsCount);
+
+                terms.Sort(indexes, SliceComparer.Instance);
+            }
+
+            public TextualFieldBuffers(IndexWriter parent)
             {
-                if (sortedTermsBuffer.Length > 0)
-                    ArrayPool<Slice>.Shared.Return(sortedTermsBuffer);
-                sortedTermsBuffer = ArrayPool<Slice>.Shared.Rent(termsCount);
+                _parent = parent;
+                Keys = ArrayPool<CompactTree.CompactKeyLookup>.Shared.Rent(BatchSize);
+                PageOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
+                PostListIds = ArrayPool<long>.Shared.Rent(BatchSize);
+                _entriesOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
```

Comment:
Do we need to clear the buffers before we start using it?

Reply:
We assume initial initialization for this process, so no need to do so.